### DISCUSSION TITLE
How can I configure the number of event loop on server?

### DISCUSSION THREAD

**Discussion by d3bt3ch** [line/armeria#6019] [Q&A] [answered]
How can I configure the number of event loop on the server? How to configure flags?

**Comment by jrhee17** [comment, marked as answer]
Hi!   If you would like to specify it globally, you can declare it via flags.
You can specify it programmatically via SPI or via JVM flags.

https://github.com/line/armeria/blob/4ba153143fb386b5dfbcc14e8200fdd26ca5a757/core/src/main/java/com/linecorp/armeria/common/Flags.java#L686-L699

Alternatively, you can specify the number of event loops per-server as well.
Obviously, values set to a server directly will have higher precedence.

https://github.com/line/armeria/blob/4ba153143fb386b5dfbcc14e8200fdd26ca5a757/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L545

**Reply by d3bt3ch** [reply]
> Hi! If you would like to specify it globally, you can declare it via flags. You can specify it programmatically via SPI or via JVM flags.
> 
> https://github.com/line/armeria/blob/4ba153143fb386b5dfbcc14e8200fdd26ca5a757/core/src/main/java/com/linecorp/armeria/common/Flags.java#L686-L699
> 
> Alternatively, you can specify the number of event loops per-server as well. Obviously, values set to a server directly will have higher precedence.
> 
> https://github.com/line/armeria/blob/4ba153143fb386b5dfbcc14e8200fdd26ca5a757/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L545

Thank you. What is then `serviceWorkerGroup` ?

**Reply by jrhee17** [reply]
By default, a worker (event loop) is responsible for the entire life cycle of a request. This includes 1) reading from a socket 2) parsing the http request 3) executing the user-specified `HttpService` (this includes grpc services, annotated services, etc..) 4) and sending back the http response.

If a `serviceWorkerGroup` is specified, user-specified `HttpService`s will be rescheduled to run from the `serviceWorkerGroup`.

**Reply by d3bt3ch** [reply]
Noted. Thanks 👍

**Reply by d3bt3ch** [reply]
Continuing in the same trail. What is `blockingTaskExecutor` and how does it affects the performance?

**Reply by ikhoon** [reply]
An event loop should not be blocked for a long time. If your service code has a blocking operation, you should use `blockingTaskExecutor`.

`blockingTaskExecutor` requires many threads and additional context switches. They could be negative impacts on the performance.

**Reply by d3bt3ch** [reply]
My service is using `CompletableFuture` to process the request. Therefore, all blocking IO tasks are within the future. Should I use `blockingTaskExecutor` in that case?
